torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+    EliminateDeadCode(graph);
+    CheckInplace(graph);
+    EliminateCommonSubexpression(graph);
+    if (!graphMustSupportVariables) {
+      // These optimizations can introduce operators like FusionGroup that
+      // do not work on variables
+      PeepholeOptimize(graph);
+      BatchMM(graph);
+      FuseGraph(graph);
+    }
+  }
+  Code & getOrCreateAutogradFallback() {
+    std::lock_guard<std::mutex> lock(compile_mutex);
+    if(autograd_fallback) {
+      return autograd_fallback;
+    }
+    auto graph_ = copy(*graph);
+    if(optimize) {
+      CreateAutodiffSubgraphs(*graph_);
+      runOptimization(graph_, /*graphMustSupportVariables=*/true);
+    }
+    autograd_fallback = Code(graph_);
+    return autograd_fallback;
+  }
+  ExecutionPlan & getOrCompile(const variable_tensor_list & inputs) {
+    // outside lock guard, to minimize the time holding the lock on the fast path
+    // ArgumentSpec even computes its hashCode here.
+    ArgumentSpec spec(autograd::GradMode::is_enabled(), inputs);
+    {
+      std::lock_guard<std::mutex> lock(compile_mutex);
+      auto it = plan_cache.find(spec);
+      if(it != plan_cache.end())
+        return it->second;
+      auto plan = compileSpec(spec);
+      auto r = plan_cache.emplace(std::move(spec), std::move(plan));
+      return r.first->second;
+    }
+  }
+  bool gradientIsPossible(const ArgumentSpec & spec) {
+    for(size_t i = 0; i < spec.size(); ++i) {
+      if(spec.tensorInfo(i).requires_grad())
+        return true;
+    }
+    return false;
+  }
+
+
+  // remove ReplaceIfUndef(v, replacement) nodes that consume inputs with 'v' if
+  // the input is defined, and 'replacement' if it is not.
+  void specializeUndef(Graph & g, const ArgumentSpec & spec) {
+    for(size_t i = 0; i < spec.size(); i++) {
+      std::vector<Value*> to_replace;
+      // do not edit in place, since it invalides uses iterator
+      for(auto u : g.inputs()[i]->uses()) {
+        if(u.user->kind() == kReplaceIfUndef) {
+          to_replace.push_back(u.user->output());
+        }
+      }
+      for(auto v : to_replace) {
+        // if it is defined, then we replace with 'v' if not,
+        // we replace with 'replacement' which is normally just a zero tensor
+        int idx = spec.tensorInfo(i).defined() ? 0 : 1;
+        v->replaceAllUsesWith(v->node()->inputs()[idx]);
+        v->node()->destroy();
+      }
+    }
+  }
+  // a + 0 -> a
+  // 0 + a -> a
+  void propagateZeros(Graph & g) {
+    for(auto it = g.nodes().begin(); it != g.nodes().end(); ++it) {
+      if(it->kind() == kadd && at::Scalar(it->t(kalpha)).toDouble() == 1.0) {
+        if(isZero(it->inputs()[0])) {
+          it->output()->replaceAllUsesWith(it->inputs()[1]);
+          it.destroyCurrent();
+        } else if(isZero(it->inputs()[1])) {
+          it->output()->replaceAllUsesWith(it->inputs()[0]);
+          it.destroyCurrent();
+        }
+      }
+    }
+  }
```

Comment:
Move to peephole (or at least mark as a TODO once we get the conservative version)?

Reply:
If we did that we have to audit peephole for checking for anything about types and sizes, because those are not specified at this point. I prefer keeping specific cleanup passes small and self-contained.